src/java.base/share/classes/java/net/URLStreamHandler.java

```diff
@@ -504,10 +504,23 @@ protected void setURL(URL u, String protocol, String host, int port,
                              String query, String ref) {
         if (this != u.handler) {
             throw new SecurityException("handler for url different from " +
-                                        "this handler");
-        } else if (host != null && u.isBuiltinStreamHandler(this)) {
-            String s = IPAddressUtil.checkHostString(host);
-            if (s != null) throw new IllegalArgumentException(s);
+                    "this handler");
+        }
+        // if early parsing, perform additional checks here rather than waiting
+        // for openConnection()
+        boolean earlyURLParsing = IPAddressUtil.earlyURLParsing();
+        boolean isBuiltInHandler = u.isBuiltinStreamHandler(this);
+        if (host != null && isBuiltInHandler) {
+            String errMsg = IPAddressUtil.checkHostString(host);
+            if (errMsg != null) throw new IllegalArgumentException(errMsg);
```

Comment:
Is it time for URLStreamHandler.setURL to document the IllegalArgumentException?

Reply:
Good point. I have logged https://bugs.openjdk.org/browse/JDK-8294948 to follow-up on this.